### PR TITLE
Decentralized Devstack(DD) prototype design decisions

### PR DESCRIPTION
**Pull request by jinder1s**
This documents some of the decisions that were made during prototyping of DD. Especially with Experiment-with-docker-interface.rst, this is a temporarily decision and will be reevaluated as we continue with prototyping.

### CODE REVIEW

**Comment by regisb** [comment]
May I suggest that you modify the title of this PR to better reflect its content? I'm not 100% clear what the changes are meant to be, so I can't suggest an appropriate title -- but it should most likely mention the devstack.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (inventhouse) [inline oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:21]: I would add "Use docker-compose commands directly _for process control_"

There are other operations I think we should support as part of a good developer experience; perhaps those should be their own ADRs, but this doc should be clear that may still be a `make` (or other) interface and that it is _only_ addressing the process control aspects.

Also this document appears to assume that _all_ our projects use docker for process control, which is not currently true.

Further, if we ever have projects that use something else or we want to smoothly migrate to something else, we would need to go back wrap everything to maintain consistency; otherwise it's not just one extra tool developers need to learn, it's N tools and the overhead of which projects use which.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (inventhouse) [inline oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:40]: Here, too, I would reiterate that there may still be makefiles that need to be maintained, but removing process control from them would lower their footprint.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (inventhouse) [inline oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:63]: Spelling "dockerk-compose"

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (inventhouse) [inline oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:61]: Especially once the case on both sides is more complete, it would be worth spelling out which aspects were ultimately most compelling in the final decision.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (inventhouse) [inline oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:52]: - Requires _all_ projects to use docker for process control
- Could increase switching cost if we ever decide to move to something else

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (jinder1s) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:8]: This decision has been made for the prototype. And will only be draft during this PR.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (jinder1s) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:14]: Please suggest better alternative wording. Otherwise, I think this clearly states need for common interface.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (kdmccormick) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:31]: > Enable a really consistent UI across projects based on different tech stacks (although we don't currently enforce this to the degree I think we should)

IMO, this is a non-goal. DD is largely built on the decisions made in OEP-41, whose purpose is to standardize back-end services on single tech stack. Furthermore, I'm generally against trying to build abstractions on top of tools that are already meant to be end-user-facing abstractions.

> Provide a "canonical example" of how to do those operations

+1, a well-curated Makefile does have a way of serving as supplemental documentation. I would list this as a pro. However, we'll need to write docs anyway, whether it be for `make` commands, `docker-compose` commands, or a mix of both.

> Reduce the onboarding learning curve by that consistency and these examples

A Makefile isn't a necessary nor sufficient condition for a consistent interface. Some Makefiles are well-curated, others  become inconsistent messes. Some docker-compose projects are non-idiomatic and confusing to use, some are well-maintained and a breeze to use if you know the four or five basic `docker-compose` commands.

I say this because I want to dispel the notion that a decision to use docker-compose directly is a decision to throw a consistent interface out the window.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (kdmccormick) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:37]: > The make wrappers can provide substantially more value than just abbreviation; I've added a few above, but there may be more I didn't think of just now.

Yes, but that doesn't change the point here that `make down|pull|up|restart` is not much shorter than `docker-compose down|pull|up|restart`, but does add a level of indirection.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (inventhouse) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:8]: I now see that line all the way down at the end; I would call it out in the title and the context sections to make this clearer.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (kdmccormick) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:57]: > A canonical interface, even if the wrappers are thin, can serve to document the "right" (or at least expected) way to do things.

If there are enough "right" and "wrong" ways to do things to the point where it's unclear how to something as simple as restart services, then I think DD has gotten too complicated.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (inventhouse) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:14]: "For Decentralized Devstack, a common, consistent, interface is necessary to enable a OpenedX Developer to add contributions to the many independent services in Open edX codebase regardless of the technology used by each project."

Comment (kdmccormick) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:14]: > regardless of the technology used by each project

This assertion is out side the scope of DD. Nothing here claims to address front-end repos, Python libraries, non-Docker legacy repos, etc.

**Review by kdmccormick — COMMENTED** [review]
Generally looks good to me :+1: Some comments...

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (kdmccormick) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:40]: > Here, too, I would reiterate that there may still be makefiles that need to be maintained, but removing process control from them would lower their footprint.

Agreed, except I've never heard the term "process control", although I can infer what you mean by it. How about "managing images and containers"?

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (inventhouse) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:14]: This statement appears to be at odds with the Application Design Pattern goal to "have a standard way to spin up a local development environment"; especially, both front-end repos and necessary services, whether legacy or not, _are_ part of our local development environment.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (inventhouse) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:31]: > MO, this is a non-goal. DD is largely built on the decisions made in OEP-41, whose purpose is to standardize back-end services on single tech stack

A usable development environment consists of more than just back-end services and that trend is accelerating

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (kdmccormick) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:14]: >> regardless of the technology used by each project

> This statement appears to be at odds with the Application Design Pattern goal to "have a standard way to spin up a local development environment";

This is true. @jinder1s , consider qualifying the Application Design Pattern goal to specify that this is for back-end services (unless, of course, you envision this applying to frontends as well).

As for legacy back-end services, I imagine that they would have to be modernized towards OEP-41 before implementing DD for them. So, I'm not trying to say that the DD design goal should not apply to say, Insights, but I think we can assume Insights would need to be modernized before the DD design goal would apply to it.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (inventhouse) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:37]: This was more about the statement "make targets do not provide much value", apparently only supported by a small amount of abbreviation.

Comment (kdmccormick) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:37]: Mm, true; my bad. "make targets do not provide much value" is a pretty strong statement.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (inventhouse) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:40]: Might be an old-school term for starting, stopping, etc. a process; "managing images and containers" may be more accurate here.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (inventhouse) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:57]: Docker itself definitely has complexities around "something as simple as restart[ing] services", with subtle differences between up and down vs. start and stop, for example.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (inventhouse) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:14]: Perhaps we should re-title the entire effort "docker development" then, if it is only about that and not about managing a working development environment, which consists of far more, increasingly especially including frontends.

I believe well-composed docker containers are an important part, but the real _need_ is for the latter.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (inventhouse) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:31]: > I say this because I want to dispel the notion that a decision to use docker-compose directly is a decision to throw a consistent interface out the window.

Also, for sure, our current interface is not as consistent as it should be right now.

Using docker-compose directly _can_ be consistent as long as _everything_ uses Docker; if some things (like frontend projects) work differently, wrappers can smooth those differences and make the developer experience more consistent.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (inventhouse) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:40]: On second thought, "managing images and containers" may be accurate for Docker-based projects; if we move everything to Docker then that's fine.

However, if we have projects that are necessary for a working development environment that are _not_ Docker-based, we need to describe the basic operations, such as launching, in a more general way.

Again, improving our use of Docker is important, but what we really, desperately, need is a much better way to manage our working development environment.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (kdmccormick) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:49]: > Using docker-compose directly can be consistent as long as everything uses Docker; if some things (like frontend projects) work differently, wrappers can smooth those differences and make the developer experience more consistent.

I think this teases at the root of our disagreement, @inventhouse .

I would prefer that front-ends standardize on one thing (e.g. NPM), back-end services standardize on another (e.g. Docker+Compose), and developers learn both `npm` and `docker-compose`. They seem like different tools for different jobs, and wrapping them may just obscure & weaken the power of the underlying tools.

But I recognize your desire to factor out the common use cases between the different tools, such as up/down/restart. What I wonder is: how would you address operations in one paradigm that have no analog in the other? For example, you can't "do database migrations" a micro-frontend. So would you implement that as a no-op for frontends? Or would that be left out of the common interface?

Comment (regisb) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:49]: Is there already a working prototype of a minimal decentralized devstack running somewhere? E.g: something that would include just the LMS, CMS, forum, an MFE and the data layer services. My feeling is that having a prototype would really help providing an immediate answer to these conversations.

As far as I am concerned, I fail to understand how such a decentralized devstack will work. For instance: would the user have to declare one or multiple docker-compose.yml files? If it is the latter, how does one automate passing the (possibly) long list of docker-compose.yml files on the CLI? How to configure the different IDAs when optional services may not be running at all? (e.g: what is the value of `COMMENTS_SERVICE_URL` when the forums may or may not be running?)

I understand that this OEP is not the right place to answer these questions; but having a working prototype would help clarify many of the technical issues (IMHO).

Comment (jinder1s) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:49]: Hi @edx/core-committer-regisb here is link to PR with working prototype: https://github.com/edx/enterprise-catalog/pull/195. It still needs to be tested, but your feedback on this would be welcome

Comment (jinder1s) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:49]: you can find documentation for it in docs folder

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (jinder1s) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:57]: I changed the contentions wording from "make targets do not provide much value" to  "these make targets do not sufficiently decrease verbosity". Which I think takes care of this contention. Feel free to unresolve this in case you disagree.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (jinder1s) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:59]: I added this in the for list above.

Comment (jinder1s) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:59]: Feel free to unresolve if you have further contentions

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (jinder1s) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:75]: I don't agree with first bullet point of this, especially with significantly decreased scope of this ADR. I've added the second bullet point to against section.

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (jinder1s) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:38]: Changing this to: "Use docker-compose commands directly for managing images and containers"

File: oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst
Comment (regisb) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:49]: This is a great example @jinder1s! I see that there are already make commands there. For instance:

    dev.migrate: # Migrates databases. Application and DB server must be up for this to work.
	docker exec -it enterprise.catalog.app bash -c 'cd /edx/app/enterprise_catalog/enterprise_catalog && make migrate'

If we were to replace this command by a docker-compose command, how would we do it? For instance we could run:

    docker-compose run --rm -f decentralized-docker-compose.yml enterprise.catalog.app make migrate

Would this be acceptable? Basically, this moves the `make` commands from the devstack to the service running inside the container.

(BTW: I strongly recommend *not* to use `docker exec`. In my experience with Tutor, this was causing major issues:

1. It requires the platform to be already running. This might not be the case because the container might be crashing because the data services are not properly configured.
2. Environment variables defined in docker-compose.yml or the .env file will not be taken into account

Sorry for the long parenthesis. If you'd like I can move this comment to the enterprise-catalog PR)

Comment (jinder1s) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:49]: for that particular question, I tried documenting here: [doc link](https://github.com/edx/enterprise-catalog/blob/msingh/spike/merge_ready/docs/decentralized_devstack_faq.rst). 

FYI, we are using .env file to toggle between decentralized devstack and legacy devstack as documented here: [toggling documentation](https://github.com/edx/enterprise-catalog/blob/msingh/spike/merge_ready/docs/decentralized_devstack_workflows.rst#toggle-between-decentralized-devstack-and-legacy-devstack)

Comment (jinder1s) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:49]: Can you elaborate on your recommendation against using docker exec? What alternative would you recommend? Are you suggesting using docker-compose exec as an alternative or wrapping docker exec in make target?

Comment (regisb) [reply oeps/oep-0005/decisions/0004-To-Make-Or-Not-To-Make.rst:49]: > Can you elaborate on your recommendation against using docker exec? What alternative would you recommend?

In my experience, `docker run --rm -it` is better than `docker exec -it`, and `docker-compose run --rm` is even better.

File: oeps/oep-0005/decisions/0004-data-dump.rst
Comment (feanil) [inline oeps/oep-0005/decisions/0004-data-dump.rst:22]:
```suggestion
In lieu of general method to provide data and in need to develop and test quickly, we've decided to place necessary lms mongo and sql data dumps in `/decentralized_devstack/data_dump` directory of the repository alongside the application code. The hope is to use this temporary data dump for testing and if this method proves promising, develop a more general solution for providing data.
```

**Review by feanil — APPROVED** [review]

**Comment by jinder1s** [comment]
Thank You @ inventhouse , @kdmccormick, and @regisb  for the feedback on this PR. After taking the feedback into account, the scope of the decisions in this PR has been significantly diminished.  There are lot of conversations that need to happen regarding interfaces for our different projects. We’ll revisit them after the prototype.

**Pull request merged**
